src/checkers/inference/InferenceQualifierHierarchy.java

```diff
@@ -224,15 +224,21 @@ public boolean isSubtype(final AnnotationMirror subtype, final AnnotationMirror
 
         final Slot subSlot   = slotMgr.getSlot(subtype);
         final Slot superSlot = slotMgr.getSlot(supertype);
-        constraintMgr.addSubtypeConstraint(subSlot, superSlot);
-
-        return true;
+        if (subSlot instanceof ConstantSlot && superSlot instanceof ConstantSlot) {
+            ConstantSlot subConstantSlot = (ConstantSlot) subSlot;
+            ConstantSlot superConstantSlot = (ConstantSlot) superSlot;
+            // Two ConstantSlot combination => immediately know the subtype relation holds or not
+            return constraintMgr.getConstraintVerifier().isSubtype(subConstantSlot, superConstantSlot);
```

Comment:
Previeously, we blindly generate subtype constraints between two slots. If there are two constant slots which are not subtype, it still generates subtype constraint and returns true. Previously, if unsatisfiable subtype constraint between two constant slot  appears, CFI doesn't exit the inference process and solver engine just ignores the unsatisfiable constraint and just disgards them. This is not correct. Since we began to directly exit the inference process if subtype constraint doesn't hold, this bug "pops up to the surface". If there is at least one VariableSlot, it makes sense to generate subtype constaint and returns true. But if there are two constant slots, it doesn't make any sense to also generate subtype constraint.

Reply:
Not really. I think `constraintMgr.addSubtypeConstraint()` should be simple as its name - only creating subtype Constraint. Also checking if subtype relation holds between two `ConstantSlot`s are business logic so I think it's better to keep it in `InferenceQualifierHieararchy`

Reply:
Is that make sense: let `constraintMgr` create a subtype constraint between two constants which are not subtype?

If it doesn't make sense, maybe adding a double check on this in `constraintMgr.addSubtypeConstraint()` would make our system a bit more robust?

Reply:
Also, if we do not let `constraintMgr` hold this checking logic, then it means for every place we need add a subtype constraint, we need write this check logic again. This might be a code duplication.

Reply:
> Is that make sense: let constraintMgr create a subtype constraint between two constants which are not subtype?

We shouldn't generate subtype constraint in this case. If such case enters `ConstraintManager`, it is a sign of having problem in one level up.

> If it doesn't make sense, maybe adding a double check on this in constraintMgr.addSubtypeConstraint() would make our system a bit more robust?

We need subtype check but for different purpose:
`ConstraintManager` checks subtype between two `ConstantSlot` is a defensive programming. Ideally, higher level should have enough business logic so that this case wouldn't happen in `ConstraintManager`. The business logic here is the `InferenceQualifierHieararchy`.
In normal typechecking mode, we check if subtype holds or not. But in inference side, we generally generate subtype constraint and returns true. But for two `ConstantSlot`s, we don't need solver to answer this - we can know this holds or not without generating any constraints. This is exactly the "business logic" for the `QualifierHierarchy` in inference side, not typechecking side.

> Also, if we do not let constraintMgr hold this checking logic, then it means for every place we need add a subtype constraint, we need write this check logic again. This might be a code duplication.

This is not correct. From top to bottom, we have `InferenceVisitor/Validator`, then `InferenceQualifierHierarchy`, and then `ConstraintManager` to add constraints. What you said only happens if you call `constraintManager.createXXXConstraint()` from `InferenceVisitor/Validator`. But this is itself wrong - `InferenceVisitor/Validator` shouldn't call `ConstraintManager` directly. `InferenceQualifierHierarchy` encapsulate all business logic in inference side, so `InferenceQualifierHierarchy` should only interact with `InferenceQualifierHierarchy`, in particular, `isSubtype()` method. So, if only you followed such practice, it doesn't introduce code duplication.

Reply:
"InferenceQualifierHierarchy encapsulate all business logic in inference side"
Not 100% correct. InferenceATF encapsulates all business logic. Some constraints are generated via the tree annotators, and the tree annotators don't always use qual hierarchy to create constraints.

Reply:
Assume we have two constant slots U, V that is not subtypes of each other. If we run inference with your changes, it crashes because qual hierarchy issues the crash --> ok no problem.

Assume U <: V  and we encounter this in qual hierarchy isSubtype --> ok no constraints generated, and it definitely type checks. I don't see how we can arrive later at the validator where U <: V doesn't type check.

Reply:
> Assume U <: V and we encounter this in qual hierarchy isSubtype --> ok no constraints generated, and it definitely type checks. I don't see how we can arrive later at the validator where U <: V doesn't type check.

Because there are also other places that call `ConstraintManager#addSubtypeConstraint()`. They can introduce the case you said. For example, `InferenceVisitor#mainIsSubtype()` can triggers this.

Reply:
Ok so if `InferenceVisitor#mainIsSubtype(@Top, @Bottom)` is called, why can't / shouldn't we just crash the program with an error trace? Why should we encode into SAT/SMT a clause for this and then delay the issue?

Reply:
`InferenceVisitor#mainIsSubtype(@Top, @Bottom)` WILL cause inference to exit, not crash(because it's not CFI or CF error. we shouldn't crash and print error stack trace) due to this line:
https://github.com/opprop/checker-framework-inference/blob/3e72f3af26a86e80f50a013f4a4fc9f2e356fc3d/src/checkers/inference/model/ConstraintManager.java#L83
So the inevitable failure is not delayed.

Dataflow also calls `isSubtype()` method to do some refinement - it can have different behaviour if the returned boolean is `true`/`false`. It's internal state, not what the programmer write. Reaching `isSubtype(@Top, @Bottom)` from dataflow doesn't mean the program is not inferrable. If you return `false`, dataflow also has its another behaviour.  

This situation happened when I ran inference on real project.